src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignedCmsTests.netcoreapp.cs

```diff
@@ -0,0 +1,498 @@
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            (X509Certificate2 cert, AsymmetricAlgorithm key) = getCertAndKey();
+            using (key)
+            {
+                CmsSigner signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, cert, key);
+                if (key is DSA)
+                {
+                    signer.IncludeOption = X509IncludeOption.EndCertOnly;
+                    signer.DigestAlgorithm = new Oid(Oids.Sha1, Oids.Sha1);
+                }
+
+                cms.ComputeSignature(signer);
+
+                byte[] ret = cms.Encode();
+                Assert.True(checkIfSignHashCalled(key));
+
+                Assert.True(ByteArrayContains(ret, content));
+
+                return (ret, cert);
+            }
+        }
+
+        private static (byte[] counterSignedData, X509Certificate2 firstSignerCert, X509Certificate2 counterSignerCert) GetCounterSignedData(byte[] content, Func<(X509Certificate2, AsymmetricAlgorithm)> getFirstSignerCertAndKey, Func<(X509Certificate2, AsymmetricAlgorithm)> getCounterSignerCertAndKey, Func<AsymmetricAlgorithm, bool> signedHashCalled, Func<AsymmetricAlgorithm, bool> signedHashCalledForCounterSignerKey)
+        {
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            (X509Certificate2 firstSignerCert, AsymmetricAlgorithm firstSignerKey) = getFirstSignerCertAndKey();
+            using (firstSignerKey)
+            {
+                CmsSigner firstSigner = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, firstSignerCert, firstSignerKey);
+                if (firstSignerKey is DSA)
+                {
+                    firstSigner.IncludeOption = X509IncludeOption.EndCertOnly;
+                    firstSigner.DigestAlgorithm = new Oid(Oids.Sha1, Oids.Sha1);
+                }
+
+                cms.ComputeSignature(firstSigner);
+                Assert.True(signedHashCalled(firstSignerKey));
+            }
+
+            byte[] signed = cms.Encode();
+            Assert.True(ByteArrayContains(signed, content));
+            SignerInfo firstSignerInfo = cms.SignerInfos[0];
+
+            (X509Certificate2 counterSignerCert, AsymmetricAlgorithm counterSignerKey) = getCounterSignerCertAndKey();
+            using (counterSignerKey)
+            {
+                CmsSigner counterSigner = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, counterSignerCert, counterSignerKey);
+                if (counterSignerKey is DSA)
+                {
+                    counterSigner.IncludeOption = X509IncludeOption.EndCertOnly;
+                    counterSigner.DigestAlgorithm = new Oid(Oids.Sha1, Oids.Sha1);
+                }
+
+                firstSignerInfo.ComputeCounterSignature(counterSigner);
+                Assert.True(signedHashCalledForCounterSignerKey(counterSignerKey));
```

Comment:
This seems like over-testing to me.  Is there a concern that this addresses which justifies the code complexity of measuring it?

If yes, maybe it could be less complex, with it being done in caller code, avoiding the callbacks?